community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/helpers/TypeSafeMathSupport.scala

```diff
@@ -219,4 +219,62 @@ trait TypeSafeMathSupport {
 
     }
   }
+
+  /**
+   * Serves for a type-safe, overflow-aware summation.
+   * That is, tries to keep the sum type as narrow as possible, widening its value if necessary.
+   * Initial result type is Int. If a Double is added, the result becomes Double.
+   * If Long is added, the result becomes Long with the possibility to widen to Double later.
+   *
+   * @note  For the summation of floating-point numbers this class encapsulates Kahan's algorithm
+   *        for error minimization (https://en.wikipedia.org/wiki/Kahan_summation_algorithm)
+   */
+  final class OverflowAwareSum {
+    import OverflowAwareSum._
+
+    private var sum: Any = 0
+    def value = sum
+
+    /**
+     * @param next  next number to add to the total sum
+     * @return      current value of the sum variable
+     * @note        sticks to integral type for as long as possible
+     */
+    def add(next: Any) = {
+      if (sum.isInstanceOf[Double] || next.isInstanceOf[Double]) {
+        sum = plus(sum, next)
+      } else if (sum.isInstanceOf[Long] || next.isInstanceOf[Long]) {
+        sum = addExact(toLong(next), toLong(sum))
+      } else if (next.isInstanceOf[Int] || sum.isInstanceOf[Int]) {
+        sum = addExact(toInt(next), toInt(sum))
+      } else {
+        sum = plus(sum, next)
+      }
+      sum
+    }
+  }
```

Comment:
This looks like it will do a lot of number boxing (please correct me if I'm misunderstanding Scala in this respect) which could be reduced if the type widening was instead expressed with subclasses of the `OverflowAwareSum`, and the `add` method returned the right subclass; itself if it is wide enough, or a new instance of the next wider subclass. This could also reduce the number of type checks you'd have to do. How does that sound?